### ISSUE TITLE
Clarification in SEDD Parameterization

### ISSUE THREAD

**Issue by yusowa0716** [kuleshov-group/bd3lms#35]
Hello,

I have a question about the sedd parameterization.

My understanding is that the goal of this function is to model the probability ratio $s\_\\theta(x\_t, t)\_y \\approx \\frac{p\_t(y)}{p\_t(x\_t)}$. I've been trying to derive the final expression from the underlying principles of the absorbing state diffusion process.

My line of reasoning is as follows:

1.  The probability of a token $x\_0$ *not* being masked at a given noise level $\\sigma(t)$ is the survival probability, which can be expressed as $p(x\_t=x\_0 | x\_0) = e^{-\\sigma(t)}$.
2.  The probability of the original token $x\_0$ transitioning to another specific token $y$ is proportional to the probability of being masked, $(1 - e^{-\\sigma(t)})$, distributed according to the model's prediction of the original data, which we can denote as $p\_\\theta(x\_0)$. This gives a transition probability of $p(x\_t=y | x\_0) = (1 - e^{-\\sigma(t)}) \\cdot \\langle p\_\\theta(x\_0), y \\rangle$.
3.  Taking the logarithm of the ratio of these probabilities, $\\log \\left( \\frac{p(x\_t=y | x\_0)}{p(x\_t=x\_0 | x\_0)} \\right)$, seems to correspond to the first part of the expression in the code, related to the model's `logits` and the `esigm1_log` term.

However, I am having trouble understanding the origin of the final term in this line of code:

```python
# From the _sedd_parameterization function
logits = logits - esigm1_log - np.log(logits.shape[-1] - 1)
```

**My specific question is:**

What is the theoretical purpose of subtracting `np.log(logits.shape[-1] - 1)`? This term does not appear in my derivation above.

Any insight or clarification on the mathematical justification for this term would be greatly appreciated.

Thank you!

**Comment by yusowa0716** [comment]
After going further, I found that it does need a normalization factor, which, if I understand correctly, should be ```-torch.logsumexp(logits)```, not ```-np.log(logits.shape[-1] - 1)```?
